MM-14442: Merge function, to enable merging of configs

cpoile opened the pull request:
#### Summary
Merge will return a new struct of the same type as base and
patch, with patch merged into base. Specifically, patch's values will be preferred
except when patch's value is `nil`.

Restrictions/guarantees:
- base or patch will not be modified
- base and patch can be pointers or values
- base and patch must be the same type
- base and patch must have no unexported types (at the moment)
- if maps or slices are different, the entire map or slice will be
  replaced (at the moment)

WIP / Request for feedback: 
- the last two points. Not sure if we wanted those yet.
- not sure where to put the merge functions, so placed it in utils. 
Happy to learn of better style or a better location
- might have went a little overboard on the tests. Wanted to make sure 
the merge could work on deeply nested structs of all types and all 
combinations of pointers and concrete types. If its useful, I'd love to turn
this into a module sometime in the future after we beat on it for awhile.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-14442

#### Checklist
- [x] Added or updated unit tests (required for all new features)

cpoile commented:
@lieut-data

lieut-data commented on utils/merge.go:
```diff
@@ -0,0 +1,131 @@
+//   - base and patch can be pointers or values
+//   - base and patch must be the same type
+//   - base and patch must have no unexported types (at the moment)
+//   - if maps or slices are different, the entire map or slice will be replaced (at the moment)
```
👍

lieut-data commented on utils/merge.go:
```diff
@@ -0,0 +1,131 @@
+// func mergeTestStruct(base, patch *testStruct) (*testStruct, error) {
+//	   ret, err := merge(base, patch)
+//	   if err != nil {
+//         return nil, err
```
nit: just a comment, but can you indent this for consistency?

cpoile replied:
I thought I had!  (I'm bothered by that kind of thing too...)
Turns out, some of those are tabs, and my editor makes it looked aligned correctly. Will replace with spaces.

lieut-data commented on config/common.go:
```diff
@@ -141,3 +142,14 @@ func (cs *commonStore) validate(cfg *model.Config) error {
+// mergeConfig merges two configs together. The receiver's values are overwritten with the patch's
+// values except when the patch's values are nil.
+func (cs *commonStore) mergeConfig(patch *model.Config) (*model.Config, error) {
+	ret, err := utils.Merge(cs, patch)
```
This should be `cs.config`, right?

cpoile replied:
absolutely, good catch, fixed

lieut-data commented on config/common.go:
```diff
@@ -141,3 +142,14 @@ func (cs *commonStore) validate(cfg *model.Config) error {
+	if err != nil {
+		return nil, err
+	}
+	retC := ret.(model.Config)
```
Seems fine to just `return &ret.(model.Config), nil` directly, if that's allowed.

lieut-data commented on utils/merge.go:
```diff
@@ -0,0 +1,131 @@
+}
+
+// mergeRec recursively merges into ret
+func mergeRec(base interface{}, patch interface{}, ret reflect.Value) (err error) {
```
Similar comment to the other review, but I feel like the following signature would be more natural for a recursive method:
```
func mergeRec(base interface{}, patch interface{}) (ret reflect.Value, error)
```

lieut-data commented on utils/merge.go:
```diff
@@ -0,0 +1,131 @@
+		}
+		pv = pv.Elem()
+	}
+	for i := 0; i < bt.NumField(); i++ {
```
`NumField` assumes we're dealing with a struct. This method might be simpler if we moved the switch on the kind below outside the loop, and recursed on non-structs as well:
```go
switch bt.Kind() {
    case reflect.Struct: {
        for i := 0; i < bt.NumField(); i++ {
            newValue, err := mergeRec(reflect.ValueOf(bt.Field(i).Interface()))
            if err != nil {
                return errors.Wrap(err, ...)
            }
            rv.Field(i).Set(newValue)
        }
    }
    case reflect.Ptr: {
        // ...
    }
    // ...
}
```

This has the nice side effect of making the high-level `Merge` function convertible to work with non-struct types as well.

lieut-data commented on utils/merge.go:
```diff
@@ -0,0 +1,131 @@
+
+// initStruct creates a new struct of type t
+func initStruct(t reflect.Type, v reflect.Value) (err error) {
+	// WIP: if we want to support structs with unexported fields, we need to do that explicitly.
```
This seems fine: looking at the high-level comments:
> //   - base and patch must have no unexported types (at the moment)

Is it true that they /must/ not have unexported types, or simply that they would be ignored during the patch operation? (The latter would be ideal.)

lieut-data commented on utils/merge_test.go:
```diff
@@ -0,0 +1,581 @@
+	return cfg
+}
+
+func TestMergeWithConfigs(t *testing.T) {
```
Can we move these tests into the `config` package instead? (I like that we're explicitly testing the merging of configs, in addition to the general case!)

lieut-data commented on utils/merge_test.go:
```diff
@@ -0,0 +1,581 @@
+
+}
+
+func NewBool(b bool) *bool          { return &b }
```
Can we avoid exporting these?

lieut-data commented on utils/merge_test.go:
```diff
@@ -0,0 +1,581 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestMergeWithStructs(t *testing.T) {
```
One thing I don't think I'm seeing by way of coverage is a guarantee that pointers inside structs aren't being copied over nor that pointers inside the base or patch structs coming in aren't being modified in some way. Is that easy to add and thus verify?

lieut-data commented on utils/merge.go:
```diff
@@ -0,0 +1,131 @@
+	return nil
+}
+
+// initStruct creates a new struct of type t
```
It's not immediately clear to me what this function accomplishes: presumably the resulting merge will take a value from either `base` or `patch`. Either both are `nil`, and thus we end up with `nil` in the result, or we copy over one of the values -- why do we need to initialize the struct at all?

lieut-data commented on utils/merge.go:
```diff
@@ -0,0 +1,147 @@
+
+// merge recursively merges patch into base and returns the new struct, ptr, slice/map, or value
+func merge(base, patch reflect.Value) (reflect.Value, bool) {
+	baseType := base.Type()
```
Worth renaming this to `commonType`? It should be obvious that both `base` and `patch` are necessarily the same types, but it might help readability when encountering `commonType` vs. reading `baseType` and wondering about `patchType`.

cpoile replied:
Sure, makes sense.

lieut-data commented on utils/merge.go:
```diff
@@ -0,0 +1,147 @@
+		}
+		return merged, true
+
+	// reflect.Chan not handled for now
```
Technically, would `default` handle `reflect.Chan`, preferring `patch` over `base` like any other value? Was the intent to fail `reflect.Chan` somehow, or leave it `nil` regardless of the incoming values?

cpoile replied:
I had trouble thinking of a reason a chan would be in a config file -- is there a use case for it? I'd be happy to add tests for it!

lieut-data commented on utils/merge.go:
```diff
@@ -0,0 +1,147 @@
+	// reflect.Chan not handled for now
+
+	default:
+		if base != patch {
```
According to https://golang.org/pkg/reflect/#Value:
> To compare two Values, compare the results of the Interface method. Using == on two Values does not compare the underlying values they represent.

We could use `.Interface()` to preserve the intended check:
```go
		if base.Interface() != patch.Interface() {
```

but I think this check is effectively redundant and you can simplify the whole block to just `return patch, true`. Since the underlying types aren't pointers at this point, either the underlying values are the same (so it doesn't matter which one is returned), or they are different and `patch` /should/ always win.

cpoile replied:
Yep, absolutely, that works perfectly.

lieut-data commented on utils/merge.go:
```diff
@@ -0,0 +1,147 @@
+//   - base and patch will not be modified
+//   - base and patch can be pointers or values
+//   - base and patch must be the same type
+//   - if slices are different, they will be merged according to the following rules:
```
I really like the flexibility to support these use cases -- that being said (and speaking selfishly ;)), merging of `*model.Config` should replace slices and maps atomically vs. merging their contents. If we want to presere these semantics, could we expose a configuration to control this?
```
type SliceMergeConfig int

const (
    ReplaceSlices                SliceMergeConfig = iota
    MergeSlicesOverwrite
)

type MapMergeConfig int

const (
    ReplaceMaps                MapMergeConfig = iota
    MergeMapsOverwrite
)

type MergeConfig struct {
    SliceMergeConfig SliceMergeConfig
    MapMergeConfig MapMergeConfig
}

func Merge(base interface{}, patch interface{}, config *MergeConfig) (interface{}, error) {
    // ...
}
```




This would also fit well into a planned extension for this library: callback functions to handle specific cases. For example, I'm presently annotating `*model.Config` with [StructTags](https://golang.org/pkg/reflect/#StructTag) to configure which fields can be modified by a restricted system admin. I could extend this config parameter with a `MergeCheckCallback func(base, patch reflect.Value, tag reflect.StructTag) bool` to effectively reuse your recursion for a customized patch case.

cpoile replied:
That sounds great, I added it verbatim -- I'm still getting a handle on the codebase, so I'm not sure if this function-specific config belongs in the merge.go file? (That's where I put it for now.) Please let me know.

lieut-data commented on utils/merge.go:
```diff
@@ -0,0 +1,147 @@
+		return mergedPtr, true
+
+	case reflect.Slice:
+		// merge with these rules:
```
Can you elaborate on when this might particular algorithm might be "useful"? I'm trying to think outside the box of just merging `*model.Config`, but I'm not sure I can imagine a case where this kind of "ovewrite" would be useful. In particular, if the slice was of type `[]*string`, we would happily replace non-`nil` values in the `base` slice with `nil` values from the `patch` slice, which runs counter to the overall idea of "merging" as such.

cpoile replied:
Yes, that makes sense. I wasn't really sure how to handle slice and map merging here. I'll take it out and have an atomic overwrite if not nil.

lieut-data commented on utils/merge.go:
```diff
@@ -0,0 +1,147 @@
+		return merged, true
+
+	case reflect.Map:
+		// for now, maps are merged in a very rudimentary way:
```
In general, I can imagine these merge semantics being more useful than the slice semantics above, but a few thoughts if we keep this:
* Should `nil` values in `patch` be ignored instead of clobbering `non-nil` values in `base`?
* If both `base` and `patch` are non-nil, it seems like we could safely clone reference values by recursing again:
```
value, ok := merge(base.MapIndex(key), patch.MapIndex(key))
if ok {
    merged.SetMapIndex(key, value)
}
```
* If one of `base` or `patch` is `nil`, I think we can still safely clone:
```
value, ok := merge(reflect.Zero(baseType), baseOrPatchWhicheverIsNotNil.MapIndex(key))
if ok {
    merged.SetMapIndex(key, value)
}
```

cpoile replied:
Good catch! I've changed it now to reflect this better behavior.

lieut-data commented on utils/merge.go:
```diff
@@ -0,0 +1,147 @@
+}
+
+// merge recursively merges patch into base and returns the new struct, ptr, slice/map, or value
+func merge(base, patch reflect.Value) (reflect.Value, bool) {
```
I'm wondering if we can eliminate the `bool` return parameter here: it seems we only use it to signal that a `nil` value was returned, but the caller had already initialized to that value anyway, so there's no loss to just assigning it again. Thoughts?

cpoile replied:
I tried and was banging my head on the wall about this. `reflect.Value` can't be nil, so you always have to return something, and the zero value of a concrete type can't be nil (of course). So I couldn't figure out how to set the pointer to nil without that extra `bool` return param.

This isn't a problem when the value is a pointer (`merge.go` line 108-109) since we check for nil value above and return the newly created (and thus nil) mergedPtr if we're setting to nil (line 100).

lieut-data commented on utils/merge_test.go:
```diff
@@ -0,0 +1,829 @@
+			[]int{1, 2, 3}, map[string]int{"key1": 1, "key2": 2},
+			simple2{30, newString("test"), []string{"test1", "test2"}},
+			&simple2{40, newString("test2"), []string{"test3", "test4", "test5"}}}
+		t2 := simple{13, 53.1, newFloat64(932.2), newInt(46), newBool(false),
```
nit: "vary" the float value here too?

cpoile replied:
see next comment

lieut-data commented on utils/merge_test.go:
```diff
@@ -0,0 +1,829 @@
+
+func TestMergeWithStructs(t *testing.T) {
+	t.Run("merge identical structs - simple", func(t *testing.T) {
+		t1 := simple{42, 42.2, newFloat64(932.2), newInt(45), newBool(true),
```
nit: I might recommend explicitly naming the fields here, as it's hard to know which are exported looking at the initialization alone.

cpoile replied:
I was thinking of just keeping these at the bottom of the test file because they are kind of 'comprehensive' -- but they are definitely too much. So I made a few more specialized tests. Let me know if there's a case that I'm missing.

lieut-data commented:
@crspeller, thanks for the mergo reminder. I had [looked](https://community.mattermost.com/core/pl/z6gydycpifnt3jt1c61kpoehge) into this /briefly/, and concluded at the time that it didn't seem to handle pointer values. Looking at this again now, I don't believe my analysis was exhaustive enough to rule it out, especially with its support for Transformers.

@cpoile, would you be interested in seeing if mergo might fit the bill? Or, if not, maybe we could fork and make any necessary tweaks? Agree with @crspeller that it's ultimately better for the community overall to leverage an existing library for something like this. Apologize for the thrashing :/

cpoile commented:
@crspeller @lieut-data Good points. I will take a look at it today and see if it does what we need it to do. Would you want me to update this PR with mergo (if it works) or start a new PR?

lieut-data commented:
If mergo fits the bill, I think it would be fine to update this PR since there's still good coverage of the explicit config merge path that we'd want to preserve. It /may/ still be the case that we want to proceed with your changes here, especially if `mergo` isn't a drop-in replacement. We could parallelize efforts to switch to `mergo` later if we could make improvements there. (Using the force a lot here!)

I'll also have to rethink my strategy in my pending PR, as I had hoped to be able to introspect the `StructTag` as we walked the `model.Config`, but I don't think mergo transformers support that...

cpoile commented:
Okay, Mergo will _almost_ work. And maybe it's fine for us. Here's where it's failing. A minimal example:
```
type evenSimpler struct {
	ES2 *evenSimpler2
}

type evenSimpler2 struct {
	S *string
}
```
And this test:
```
t.Run("patch nils are ignored - evenSimpler", func(t *testing.T) {
	t1 := evenSimpler{&evenSimpler2{newString("base")}}
	t2 := evenSimpler{&evenSimpler2{nil}}
	expected := evenSimpler{&evenSimpler2{newString("base")}}

	merged, err := mergeEvenSimpler(t1, t2)
	require.NoError(t, err)

	assert.Equal(t, expected, *merged)
})
```
`*merged.ES2.S` equals `nil` -- it's overwritten by patch. 
When the `nil` value is in a reference field in a struct, or when the `nil` is in a reference field in a non-reference embedded struct, it works. E.g.:
```
type t struct {
    S *string
}
```
or:
```
type t struct {
    T2 t2
}
type t2 struct {
    S *string
}
```
Those both work -- nils in the patch won't overwrite the base.

Now, at first I thought this would't matter... Except that we have `map[string]*PluginState`. So I tested that:
```
type mapPtrState struct {
	MP map[string]*state
}
type state struct {
	Enable bool
}

```
with:
```
t.Run("patch nils are ignored - mapPtrState", func(t *testing.T) {
	t1 := mapPtrState{map[string]*state{"base key": {true}}}
	t2 := mapPtrState{map[string]*state{"base key": nil}}
	expected := mapPtrState{map[string]*state{"base key": {true}}}

	merged, err := mergeMapPtrState(t1, t2)
	require.NoError(t, err)

	assert.Equal(t, expected, *merged)
})
```
And it failed. A `nil` patch key will overwrite an existing base key.
 Unless you know of other places where this will bite?

lieut-data commented:
@cpoile, thanks for the detailed breakdown. I had a chance to sync up with @crspeller, and the two steps forward are (in priority order):
* Investigate forking mergo and making it do what we want, optionally submitting a general purpose patch upstream. This might be complicated by the fact that they presumably have a stable API so we'd end up with a long-term fork instead.
* Adopt your code but make it laser-focussed on merging configs (i.e. drop the special slice and map handling) so as to avoid trying to solve a "general purpose" problem. Maybe it even lives in `config` to avoid ambiguity.

You've contributed a ton in this issue (and others!) already, and I don't want to burden you with the above investigation... unless you're interested in pursuing this further! Let's discuss and figure out the next steps.

cpoile commented:
I'd be happy to continue to work on it -- I'd hate to leave it unfinished. Right now the revised PR code passes all the tests (including the ones that mergo barfs on). I'll spend some time to try to figure out if I can patch it; even though I hate to not use the work, it wasn't wasted effort from my POV, and for MM it makes more sense to use a battle-tested package (as @crspeller points out). :)

cpoile commented:
Hi @lieut-data @crspeller, here is the newer version. I spent a bit too long trying to make mergo work. I thought I had it working last night, but then I thought I'd make some changes to some nested pointers. And that opened a can of worms. At the moment, mergo is:
- copying nested pointers (or at least pointers/maps/slices that is itself inside a pointer to a map/slice/struct)
- treats `*bool(nil)` the same as `*bool(false)` when deciding to merge. Which means if we have a plugin setting set to `false` it will not override a `true`, unless we explicitly tell it to override even when the src (`patch` in our lingo) is nil -- which is explicitly not what we want in a configuration file merger.
- along with other weird behavior, like a `struct{*struct2{*string}}` is overwritten by a `struct{*struct2{nil}}` 
After a while of trying to fix it, I realized that I was turning mergo into what we already have, so I just stopped.

Main changes here:
- the merge function is much simpler and focused now, thanks to Jesse's suggestions. Maps/slices are completely overwritten if patch is not nil. 
- nested pointers/maps/slices are completely cloned now
- 54 new tests gained from trying to fix mergo
- simplified and more specific tests. I kept the omnibus huge nested cases at the very end, just in case. I like them because they sometimes would catch something, and then I added a minimum reproducible test to isolate that particular issue, then fixed it.

Please let me know what can be improved and I'll get on it. Thanks for your eyes on this!

lieut-data commented on utils/merge_test.go:
```diff
@@ -0,0 +1,1648 @@
+
+}
+
+func mergeSimple(base, patch simple) (*simple, error) {
```
Heh, this is proof of why Go needs /something/ for generics. :)

cpoile replied:
No kidding. But, I'm starting to appreciate that here I get more control, and it's perfectly clear what's happening. :)

lieut-data commented on utils/merge.go:
```diff
@@ -0,0 +1,165 @@
+//   - if slices are different, this rule applies:
+//       - if patch is not nil, overwrite the base slice.
+//       - otherwise, keep the base slice
+//   - if maps are different, they will be merged according to the following rules:
```
I think this documentation block needs to be updated to reflect the simpler semantics.

cpoile replied:
Done. I had updated the one below, but missed this, thanks!

lieut-data commented on config/common.go:
```diff
@@ -4,6 +4,7 @@
 package config
 
 import (
+	"github.com/mattermost/mattermost-server/utils"
```
nit: Can you apply [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports) to this file? (We should really make this part of the CI). The almost-idiomatic way of ordering imports is something like:
```
<stdlib>

<external to this repository>

<internal to this repository>
```

`goimports` is basically `go fmt` + fixing that ordering above. I have it integrated into my editor to make it automatic.

cpoile replied:
Awesome, mine is sent up to do that now too.

lieut-data commented:
Thanks again, @cpoile :)
